pkg/datamover/utils.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2023 The Kanister Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datamover
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"os"
+
+	"github.com/pkg/errors"
+
+	"github.com/kanisterio/kanister/pkg/field"
+	"github.com/kanisterio/kanister/pkg/kopia"
+	kopiacmd "github.com/kanisterio/kanister/pkg/kopia/command"
+	"github.com/kanisterio/kanister/pkg/kopia/repository"
+	"github.com/kanisterio/kanister/pkg/kopia/snapshot"
+	"github.com/kanisterio/kanister/pkg/location"
+	"github.com/kanisterio/kanister/pkg/log"
+	"github.com/kanisterio/kanister/pkg/output"
+	"github.com/kanisterio/kanister/pkg/param"
+)
+
+const (
+	usePipeParam = `-`
+)
+
+func targetWriter(target string) (io.Writer, error) {
+	if target != usePipeParam {
+		return os.OpenFile(target, os.O_RDWR|os.O_CREATE, 0755)
+	}
+	return os.Stdout, nil
+}
+
+func locationPull(ctx context.Context, p *param.Profile, path string, target io.Writer) error {
+	return location.Read(ctx, target, *p, path)
+}
+
+// kopiaLocationPull pulls the data from a kopia snapshot into the given target
+func kopiaLocationPull(ctx context.Context, backupID, path, targetPath, password string) error {
+	switch targetPath {
+	case usePipeParam:
+		return snapshot.Read(ctx, os.Stdout, backupID, path, password)
+	default:
+		return snapshot.ReadFile(ctx, backupID, targetPath, password)
+	}
+}
+
+// connectToKopiaServer connects to the kopia server with given creds
+func connectToKopiaServer(ctx context.Context, kp *param.Profile) error {
+	contentCacheSize := kopia.GetDataStoreGeneralContentCacheSize(kp.Credential.KopiaServerSecret.ConnectOptions)
+	metadataCacheSize := kopia.GetDataStoreGeneralMetadataCacheSize(kp.Credential.KopiaServerSecret.ConnectOptions)
+	return repository.ConnectToAPIServer(
+		ctx,
+		kp.Credential.KopiaServerSecret.Cert,
+		kp.Credential.KopiaServerSecret.Password,
+		kp.Credential.KopiaServerSecret.Hostname,
+		kp.Location.Endpoint,
+		kp.Credential.KopiaServerSecret.Username,
+		contentCacheSize,
+		metadataCacheSize,
+	)
+}
+
+// connectToKopiaRepositoryServer connects to the kopia server with given repository server CR
+func connectToKopiaRepositoryServer(ctx context.Context, rs *param.RepositoryServer) (string, error) {
+	contentCacheMB, metadataCacheMB := kopiacmd.GetCacheSizeSettingsForSnapshot()
```

Comment:
we are going to take the cache settings from the CR. Check https://github.com/kanisterio/kanister/pull/1918/files#diff-5265b02a94d73bb8cf15ecfce880d68a7b1bb98cabf584800e93fe3ed7bb65fb . Can be done in a different PR, but lets keep a note of this

Reply:
Yeah Noted !

Reply:
@r4rajat since we have merged cache settings PR, can we use the these settings from repository server PR, if its not provided there, we should use the default one. WDYT?

Reply:
Done @kale-amruta , added it in RepositoryServer templating params